cilium/CFP-30984-dns-proxy-ha.md

```diff
@@ -0,0 +1,151 @@
+# CFP-30984: toFQDN DNS proxy HA
+
+**SIG: SIG-POLICY**
+
+**Begin Design Discussion:** 2024-04-08
+
+**Cilium Release:** 1.17
+
+**Authors:** Hemanth Malla <hemanth.malla@datadoghq.com>
+
+## Summary
+
+Cilium agent uses a proxy to intercept all DNS queries and obtain necessary information for enforcing toFQDN network policies. However, the lifecycle of this proxy is coupled with cilium agent. When an endpoint has a toFQDN network policy in place, cilium installs a redirect to capture all DNS traffic. So, when the agent is unavailable all DNS requests time out, including when DNS name to IP address mappings are already in place for this name.DNS policy unload on shutdown can be enabled on agent, but it works only when L7 policy is set to * and agent is shutdown gracefully.
+
+This CFP introduces a standalone DNS proxy that can be run alongside cilium agent which should eliminate the hard dependency on cilium agent atleast for names that are already resolved. 
```

Comment:
What does the "atleast for names that are already resolved" part here mean?

Are you assuming this is a caching proxy or a passthrough proxy? Cilium currently only ever does passthrough AFAIK, and changing that could have implications on things like state coherency.

Reply:
I think what he meant here is, with current dnsproxy if cilium agent down, then dns resolution fails and that impacts network connectivity. With Standalone dnsproxy, even if cilium agent is down, dns requests could be resolved and if it resolves to same IP, then existing bpf policy if any would be matched and connectivity works.

Reply:
Ah. I understand how this is ambiguous. Will update this with "eliminate hard dependency for names that already have policy map entries in place". I do not intend to change the pass through aspect of the proxy for DNS, but if we are not accessing bpf maps directly for EP and identity info, then we would need to cache the mappings in SDP.